plugin/src/main/java/net/aufdemrand/denizen/utilities/AdvancementHelper.java

```diff
@@ -0,0 +1,163 @@
+            Bukkit.getUnsafe().loadAdvancement(id, getJson());
+        }
+        catch (IllegalArgumentException e) {
+            dB.echoError("Error registering advancement!");
+        }
+    }
+
+    /**
+     * Removes the advancement
+     */
+    private void remove() {
+        Bukkit.getUnsafe().removeAdvancement(id);
+        Bukkit.getServer().reloadData();
+    }
+
+    /**
+     * Grants & displays this custom advancement by fulfilling its criteria
+     * @param players List of players
+     */
+    private void grant(Collection<? extends Player> players) {
+        Advancement advancement = Bukkit.getAdvancement(id);
+        AdvancementProgress progress;
+        for (Player player : players) {
+            progress = player.getAdvancementProgress(advancement);
+            if (!progress.isDone())	{
+                for (String criteria : progress.getRemainingCriteria())	{
+                    progress.awardCriteria(criteria);
+                }
+            }
+        }
+    }
+
+    /**
+     * Revokes the granted criteria for this advancement
+     * @param players List of players
+     */
+    private void revoke(Collection<? extends Player> players) {
+        Advancement advancement = Bukkit.getAdvancement(id);
+        AdvancementProgress progress;
+        for (Player player : players) {
+            progress = player.getAdvancementProgress(advancement);
+            if (progress.isDone()) {
+                for (String criteria : progress.getAwardedCriteria()) {
+                    progress.revokeCriteria(criteria);
+                }
+            }
+        }
+    }
+
+    /**
+     * Constructs a valid Advancement JSON String
+     * See: https://minecraft.gamepedia.com/Advancements#JSON_Format
+     * TODO: Implement unused portions in Advancement Script Containers
+     * @return Formatted Advancement JSON String
+     */
+    private String getJson()	{
+
+        JsonObject json = new JsonObject();
+
+        JsonObject icon = new JsonObject();
+        icon.addProperty("item", this.icon);
+
+        JsonObject display = new JsonObject();
+        display.add("icon", icon);
+        display.addProperty("title", this.title);
+        display.addProperty("description", this.description);
+        display.addProperty("background", "minecraft:textures/gui/advancements/backgrounds/adventure.png");
+        display.addProperty("frame", this.frame);
+        display.addProperty("announce_to_chat", announce);
+        display.addProperty("show_toast", toast);
+        display.addProperty("hidden", true);
+
+        JsonObject criteria = new JsonObject();
+        JsonObject trigger = new JsonObject();
+
+        trigger.addProperty("trigger", "minecraft:impossible");
+        criteria.add("impossible", trigger);
+
+        json.add("criteria", criteria);
+        json.add("display", display);
+
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+        return gson.toJson(json);
+
+    }
+}
```

Comment:
What the hell is this file? It's full of format errors and weird semidocumentation.